### ISSUE TITLE
Introduce an API for referring to types available in a particular semantic context

### ISSUE THREAD

**Issue by malinthar** [ballerina-platform/ballerina-lang#33121]
**Description:**

```ballerina
type ErrorDetail1 record {|
    string cause;
    int errorCode;
|};

type MyErr error<ErrorDetail1>;

public function main() {
    MyErr myerr = error(<cursor>)
}
```
An error constructor requires a mandatory positional arg (error message). Currently, there is not a way to get the type descriptor of this parameter. Following is a capture of the error symbol of the above source code.
![error](https://user-images.githubusercontent.com/35211477/136366439-9e0a063e-e01f-4cc2-83e7-062110bf38cb.png)

We need to capture the expected type at the cursor position in order to check for the assignability to  provide a sorted completion items list.

**Comment by pubudu91** [comment]
The more generic problem here is that there is no way in the Semantic API to refer to builtin types. e.g., referring to `string`, `int` etc. type symbols

This ability to reference these types is vital to do queries like "is `T` a sub type of `string`?". We need to come up with an API for this.

**Comment by pubudu91** [comment]
So something like the following is what I had in mind.

![types](https://user-images.githubusercontent.com/6260009/145400724-bd39f2b9-0894-4bbf-af6f-b4b4356c5bde.jpg)

The main requirement is one should be able to refer to the types available in a particular semantic context in order to be able to do various analysis tasks. This came up a lot in compiler plugin development related stuff. For now, what they have done is manually verify types using name, module ID etc. Instead, they should be able to use the `subtypeOf()` API in `TypeSymbol`. To do that, they need access to the types. 

A next step in this API would be to provide support for creating type symbols.

**Comment by pubudu91** [comment]
The design doc for this: https://docs.google.com/document/d/15I164NKdPX0xu40Nxiy6wL_eJg2rvG2YFtUNDdcKgZ8/edit?usp=sharing

**Comment by mohanvive** [comment]
Initial implementation is merged with https://github.com/ballerina-platform/ballerina-lang/pull/35353 . Next set of improvements is in progress.

**Comment by sanjana** [comment]
> Initial implementation is merged with #35353 . Next set of improvements are in progress.

The improvements are tracked as subtasks of #35415. 
Closing this since the initial requirement was resolved by #35353.
